examples/solvers/bregman_tv_tomography.py

```diff
@@ -0,0 +1,125 @@
+Szymkat M. Advances in Mathematical Modeling, Optimization and Optimal Control.
+Springer, 2016, pp 3--33.
+https://doi.org/10.1007/978-3-319-30785-5_2
+https://arxiv.org/abs/1505.05191
+"""
+
+import odl
+import numpy as np
+
+# Reconstruction space
+reco_space = odl.uniform_discr(min_pt=[-20, -20], max_pt=[20, 20],
+                               shape=[128, 128], dtype='float32')
+
+# Make a parallel beam geometry with flat detector, and create ray transform
+geometry = odl.tomo.parallel_beam_geometry(reco_space,  num_angles=100)
+ray_trafo = odl.tomo.RayTransform(reco_space, geometry)
+
+# Create phantom, forward project to create sinograms, and add 10% noise
+discr_phantom = odl.phantom.shepp_logan(reco_space, modified=True)
+noise_free_data = ray_trafo(discr_phantom)
+noise = odl.phantom.white_noise(ray_trafo.range)
+noise *= 0.10 / noise.norm() * noise_free_data.norm()
+data = noise_free_data + noise
+
+# Components for variational problem: l2-squared data matching and isotropic
+# TV-regularization
+l2_norm = 0.5 * odl.solvers.L2NormSquared(ray_trafo.range).translated(data)
+gradient = odl.Gradient(reco_space)
+reg_param = 0.3
+l12_norm = reg_param * odl.solvers.GroupL1Norm(gradient.range)
+
+# Assemble functionals and operators for the optimization algorithm
+f = odl.solvers.ZeroFunctional(reco_space)  # No f functional used, set to zero
+g = [l2_norm, l12_norm]
+L = [ray_trafo, gradient]
+
+# Estimated operator norms, which are used to ensure that we fulfill the
+# convergence criteria of the optimization algorithm
+ray_trafo_norm = odl.power_method_opnorm(ray_trafo, maxiter=20)
+gradient_norm = odl.power_method_opnorm(gradient, maxiter=20)
+
+# Parameters for the optimization method; tuned in order to reduce the number
+# of inner iterations needed to solve the first step in the Bregman iterations
+niter_inner = 200
+tau = 0.01  # Step size for the primal variable
+sigma_ray_trafo = 45.0 / ray_trafo_norm ** 2  # Step size for dual variable
+sigma_gradient = 45.0 / gradient_norm ** 2  # Step size for dual variable
+sigma = [sigma_ray_trafo, sigma_gradient]
+
+# The reconstruction looks nice after about 5 outer iterations; set total
+# number of outer iterations to 7 to show what happens if one does to many
+niter_bregman = 7
+
+# Create initial guess and initial subgradient
+x = reco_space.zero()
+p = reco_space.zero()
+
+# This defines the outer Bregman iterations
+for breg_iter in range(niter_bregman):
+    print('Outer Bregman iteration: {}'.format(breg_iter))
+
+    # Create the affine part of the Bregman functional
+    constant = l12_norm(gradient(x))
+    linear_part = reg_param * odl.solvers.QuadraticForm(vector=-p,
+                                                        constant=constant)
+
+    callback_inner = odl.solvers.CallbackPrintIteration(step=50)
+
+    # Inner iterations; x is updated in-place with the consecutive iterates
+    odl.solvers.forward_backward_pd(
```

Comment:
Without the restrictions of ODL, I would relate `h` with the data fit. The update of the subgradient includes the gradient of the data fit at the last iterate. Thus, one could recycle these computations. It would be good if ODL had the option to do this. In particular for real size problems. This strategy would also be crucial for the linearized Bregman iterations.

Reply:
That will all be possible when #1400 lands.

Reply:
As of now, one can wrap both the data matching and the linear part of the Bregman functional in a `FunctionalQuadraticPerturb`, and pass all of that as the smooth part of the solver (i.e., as `h`). I guess it is something like that you were looking for?

However, before we have something like #1400 in place it will be hard (impossible?) to reuse the gradients computed by the solver.